### PR TITLE
Board templates

### PR DESCRIPTION
**Pull request by vrialland**

### CODE REVIEW

File: kansha/alembic/versions/2b0edcfa57b4_add_templates.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ def upgrade():
     services.register('mail_sender', DummyMailSender())
 
     bm = BoardsManager('', '', '', {}, None, services)
+    bm.create_template_empty()
```
Comment (nagareproject) [inline kansha/alembic/versions/2b0edcfa57b4_add_templates.py]: Use name with meanings, no names on 1 or 2 characters

File: kansha/board/boardsmanager.py [hunk 2]
```diff
@@ -14,6 +14,7 @@
 from glob import glob
 
 from nagare import database
+from nagare.i18n import _
```
Comment (nagareproject) [inline kansha/board/boardsmanager.py]: imports order

File: kansha/board/boardsmanager.py [hunk 3]
```diff
@@ -59,15 +60,23 @@ def _get_board(self, data_board):
         return self._services(Board, data_board.id, self.app_title, self.app_banner, self.theme,
                               self.card_extensions, self.search_engine)
 
+    def create_template_empty(self):
+        '''Get a default empty template'''
+        board = DataBoard(title=u'Empty board', is_template=True)
```
Comment (nagareproject) [inline kansha/board/boardsmanager.py]: Important: a component can use the data part of an other component. Component uses an other component thru its public API exposed from its logic part. So, here, create a `Board`, not a `DataBoard`

File: kansha/board/boardsmanager.py [hunk 4]
```diff
@@ -59,15 +60,23 @@ def _get_board(self, data_board):
...
+        board.visibility = BOARD_PUBLIC
```
Comment (nagareproject) [inline kansha/board/boardsmanager.py]: Could it be possible to derivate `TemplateBoard` from `Board`? So `is_template` and the default `visibility` could be set in `TemplateBoard`

File: kansha/board/boardsmanager.py [hunk 5]
```diff
@@ -59,15 +60,23 @@ def _get_board(self, data_board):
...
+        for i, (title, color) in enumerate(DEFAULT_LABELS):
+            board.labels.append(DataLabel(title=title, color=color, index=i))
```
Comment (nagareproject) [inline kansha/board/boardsmanager.py]: Demeter's law: this component can't access the `labels` of `board`. It can only call method on `board`

Comment (nagareproject) [inline kansha/board/boardsmanager.py]: The `index` property must be managed internally by the `labels`

File: kansha/board/boardsmanager.py [hunk 6]
```diff
@@ -59,15 +60,23 @@ def _get_board(self, data_board):
...
+        database.session.flush()
+        return self._get_board(board)
```
Comment (nagareproject) [inline kansha/board/boardsmanager.py]: These 2 lines are not needed if (when) you already are working with a true `Board` component

File: kansha/board/boardsmanager.py [hunk 7]
```diff
@@ -59,15 +60,23 @@ def _get_board(self, data_board):
...
+
     def create_template_todo(self):
         '''Get a default Todo template'''
         board = DataBoard(title=u'Todo', is_template=True)
         board.visibility = BOARD_PUBLIC
-        for title, index in [(u'To Do', 0), (u'Doing', 1), (u'Done', 2)]:
+        for index, title in enumerate((u'To Do', u'Doing', u'Done')):
```
Comment (nagareproject) [inline kansha/board/boardsmanager.py]: The `index` property must be managed internally by the board

File: kansha/board/boardsmanager.py [hunk 8]
```diff
@@ -59,15 +60,23 @@ def _get_board(self, data_board):
...
+        board = DataBoard(title=u'Empty board', is_template=True)
+        board.visibility = BOARD_PUBLIC
+        for i, (title, color) in enumerate(DEFAULT_LABELS):
+            board.labels.append(DataLabel(title=title, color=color, index=i))
+        database.session.flush()
+        return self._get_board(board)
+
     def create_template_todo(self):
         '''Get a default Todo template'''
         board = DataBoard(title=u'Todo', is_template=True)
```
Comment (nagareproject) [inline kansha/board/boardsmanager.py]: Idem: create a `Board` component

File: kansha/board/boardsmanager.py [hunk 9]
```diff
@@ -59,15 +60,23 @@ def _get_board(self, data_board):
         return self._services(Board, data_board.id, self.app_title, self.app_banner, self.theme,
                               self.card_extensions, self.search_engine)
 
+    def create_template_empty(self):
```
Comment (nagareproject) [inline kansha/board/boardsmanager.py]: Note: this method will have to be removed when the importation of board will be available

File: kansha/board/boardsmanager.py [hunk 10]
```diff
@@ -59,15 +60,23 @@ def _get_board(self, data_board):
...
+        '''Get a default empty template'''
+        board = DataBoard(title=u'Empty board', is_template=True)
+        board.visibility = BOARD_PUBLIC
+        for i, (title, color) in enumerate(DEFAULT_LABELS):
+            board.labels.append(DataLabel(title=title, color=color, index=i))
+        database.session.flush()
+        return self._get_board(board)
+
     def create_template_todo(self):
```
Comment (nagareproject) [inline kansha/board/boardsmanager.py]: Note: this method will have to be removed when the importation of board will be available

File: kansha/board/boardsmanager.py [hunk 11]
```diff
@@ -59,15 +60,23 @@ def _get_board(self, data_board):
         return self._services(Board, data_board.id, self.app_title, self.app_banner, self.theme,
                               self.card_extensions, self.search_engine)
 
+    def create_template_empty(self):
+        '''Get a default empty template'''
+        board = DataBoard(title=u'Empty board', is_template=True)
+        board.visibility = BOARD_PUBLIC
+        for i, (title, color) in enumerate(DEFAULT_LABELS):
```
Comment (nagareproject) [inline kansha/board/boardsmanager.py]: To me, an empty board hasn't even predefined labels

File: kansha/board/comp.py [hunk 12]
```diff
@@ -161,6 +163,10 @@ def many_user_render(h, number):
 
         self.must_reload_search = False
 
+    def reload_save_template_comp(self):
```
Comment (nagareproject) [inline kansha/board/comp.py]: Why is such method needed?

File: kansha/board/comp.py [hunk 13]
```diff
@@ -161,6 +163,10 @@ def many_user_render(h, number):
...
+        self.save_template_comp = component.Component(SaveTemplateTask(self))
+        self.save_template_comp.on_answer(lambda v: self.reload_save_template_comp())
+
     def copy(self, owner, additional_data):
         new_data = self.data.copy(None)
         new_obj = self._services(Board, new_data.id, self.app_title, self.app_banner, self.theme, self.card_extensions, self.search_engine, load_data=False)
```
Comment (nagareproject) [inline kansha/board/comp.py]: Semantic naming: `new_obj` -> `new_board`

File: kansha/board/comp.py [hunk 14]
```diff
@@ -168,23 +174,26 @@ def copy(self, owner, additional_data):
         additional_data['author'] = owner
 
         cols = [col() for col in self.columns if not col().is_archive]
```
Comment (nagareproject) [inline kansha/board/comp.py]: Note: when a default archive column will be always created, this `is_archive` stuff will have to be removed

File: kansha/board/comp.py [hunk 15]
```diff
@@ -168,23 +174,26 @@ def copy(self, owner, additional_data):
...
-        for index, column in enumerate(cols):
+        index = 0
+        for column in cols:
             new_col = column.copy(new_obj, additional_data)
```
Comment (nagareproject) [inline kansha/board/comp.py]: Let an 'ordering_collection' of SQLAlchemy manage the index

File: kansha/board/comp.py [hunk 16]
```diff
@@ -161,6 +163,10 @@ def many_user_render(h, number):
 
         self.must_reload_search = False
 
+    def reload_save_template_comp(self):
+        self.save_template_comp = component.Component(SaveTemplateTask(self))
+        self.save_template_comp.on_answer(lambda v: self.reload_save_template_comp())
+
     def copy(self, owner, additional_data):
```
Comment (nagareproject) [inline kansha/board/comp.py]: What is this `additional_data`?

File: kansha/board/comp.py [hunk 17]
```diff
@@ -168,23 +174,26 @@ def copy(self, owner, additional_data):
...
         for lbl in self.labels:
             new_obj.labels.append(lbl.copy(new_obj, additional_data))
 
         return new_obj
 
-    def save_as_template(self, shared):
+    def save_as_template(self, title, description, shared):
         user = security.get_user()
         template = self.copy(user, {})
+        template.data.title = title
```
Comment (nagareproject) [inline kansha/board/comp.py]: A component can't access the `data` of another one

File: kansha/board/boardsmanager.py (hunk 11 above)
Comment (vrialland) [inline kansha/board/boardsmanager.py]: At this point, there is no way to create a label, so you won't be able to use labels in that case.

File: kansha/board/comp.py (hunk 16 above)
Comment (vrialland) [inline kansha/board/comp.py]: Some informations need to be propagated through the components tree, for example labels are created at the board level and must be available to cards as they use it. These are stored in this dict to be used later.

File: kansha/board/save_template.py [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+#--
+
+from nagare import component, editor, presentation, validator as nagare_validator
+from nagare.i18n import _
+
+from kansha import validator
+from kansha.toolbox import remote
+
+
+class SaveTemplateEditor(editor.Editor):
```
Comment (nagareproject) [inline kansha/board/save_template.py]: Don't inherit from `editor.Editor` if you don't use its core API (change `if self.is_validated(('title', 'shared'))` by `if title.error is None and shared.error is None`

File: kansha/board/save_template.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+    def cancel(self, comp):
+        comp.answer(None)
+
+    def commit(self, comp):
+        if self.is_validated(('title', 'shared')):
+            comp.answer((self.title.value, self.description.value, self.shared.value))
+
+    def close(self, comp):
+        comp.answer()
+        return 'YAHOO.kansha.app.hideOverlay()'
```
Comment (nagareproject) [inline kansha/board/save_template.py]: Don't mix logic and presentation

File: kansha/board/save_template.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
+# -*- coding:utf-8 -*-
```
Comment (nagareproject) [inline kansha/board/save_template.py]: This module must move to `Template` or `TemplateManager`

File: kansha/board/save_template.py [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+        comp.answer()
+        return 'YAHOO.kansha.app.hideOverlay()'
+
+
+@presentation.render_for(SaveTemplateEditor)
+def render_SaveTemplateEditor(self, h, comp, *args):
+    with h.form(class_='description-form'):
+        with h.div:
+            id_ = h.generate_id()
+            h << h.label(_(u'Title'), for_=id_)
```
Comment (nagareproject) [inline kansha/board/save_template.py]: Don't use unicode strings when it's not needed

File: kansha/board/save_template.py [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+            h << h.label(_(u'Save for'), for_=id_)
+            with h.select(id_=id_).error(self.shared.error).action(self.shared):
+                h << h.option(_(u'Me only'), value='').selected(not self.shared())
+                h << h.option(_(u'All users'), value='on').selected(self.shared())
+
+        with h.div(class_='buttons'):
+            h << h.button(_(u'Save'), class_='btn btn-primary', type='submit').action(self.commit, comp)
+            h << ' '
+            h << h.button(_('Cancel'), class_='btn').action(remote.Action(lambda comp=comp: self.close(comp)))
+    return h.root
```
Comment (nagareproject) [inline kansha/board/save_template.py]:
```python
remote.Action(lambda comp=comp: self.close(comp))
```
->
```python
remote.Action(lambda: self.close(comp))
```

File: kansha/board/save_template.py [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+
+
+@presentation.render_for(SaveTemplateEditor, 'saved')
+def render_SaveTemplateEditor_saved(self, h, comp, *args):
+    close = remote.Action(lambda comp=comp: self.close(comp))
```
Comment (nagareproject) [inline kansha/board/save_template.py]:
```python
remote.Action(lambda comp=comp: self.close(comp))
```
->
```python
remote.Action(lambda: self.close(comp))
```

File: kansha/board/save_template.py [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+    h << h.script('''YAHOO.kansha.app.onHideOverlay(function() { %s; });''' % close.generate_action(1, h))
```
Comment (nagareproject) [inline kansha/board/save_template.py]: Don't use the low level `generate_action()`

File: kansha/populate.py [hunk 25]
```diff
@@ -33,6 +33,7 @@ def populate():
     services.register('mail_sender', DummyMailSender())
 
     bm = BoardsManager('', '', '', {}, None, services)
```
Comment (nagareproject) [inline kansha/populate.py]: Use name with meanings, no names on 1 or 2 characters

File: kansha/board/comp.py (hunk 16 above)
Comment (Net-ng) [inline kansha/board/comp.py]: > Le 2 déc. 2015 à 17:09, Vincent Rialland <notifications@github.com> a écrit :
> 
> In kansha/board/comp.py:
> 
> > @@ -161,6 +163,10 @@ def many_user_render(h, number):
> >  
> >          self.must_reload_search = False
> >  
> > +    def reload_save_template_comp(self):
> > +        self.save_template_comp = component.Component(SaveTemplateTask(self))
> > +        self.save_template_comp.on_answer(lambda v: self.reload_save_template_comp())
> > +
> >      def copy(self, owner, additional_data):
> 
> Some informations need to be propagated through the components tree, for example labels are created at the board level and must be available to cards as they use it.

This copy of informations will disappear in the near future

> These are stored in this dict to be used later.

If you really need to propagate informations down to the cards, explicitly named
them (i.e: `labels` and not `additional_data`)

**Pull request closed without merging**